lib/plugins/aws/invokeLocal/index.js

```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
         this.options.context);
     }
 
-    throw new this.serverless.classes
-      .Error('You can only invoke Node.js, Python, Java & Ruby functions locally.');
+    return this.invokeLocalDocker();
+  }
+
+
+  checkDockerImage() {
+    const runtime = this.options.functionObj.runtime
+      || this.serverless.service.provider.runtime
+      || 'nodejs4.3';
+
+    return new BbPromise((resolve, reject) => {
+      const docker = spawn('docker', ['images', '-q', `lambci/lambda:${runtime}`]);
+      let stdout = '';
+      docker.stdout.on('data', (buf) => { stdout += buf.toString(); });
+      docker.on('close', error => (error ? reject(error) : resolve(Boolean(stdout.trim()))));
+    });
+  }
+
+  pullDockerImage() {
+    const runtime = this.options.functionObj.runtime
+      || this.serverless.service.provider.runtime
+      || 'nodejs4.3';
+
+    this.serverless.cli.log('Downloading base Docker image...');
+
+    return new BbPromise((resolve, reject) => {
+      const docker = spawn('docker', ['pull', `lambci/lambda:${runtime}`]);
+      docker.on('close', error => (error ? reject(error) : resolve()));
+    });
+  }
+
+  getLayerPaths() {
+    const layers = _.mapKeys(
+      this.serverless.service.layers,
+      (value, key) => this.provider.naming.getLambdaLayerLogicalId(key)
+    );
+
+    return BbPromise.all(
+      (this.options.functionObj.layers || this.serverless.service.provider.layers || [])
+        .map(layer => {
+          if (layer.Ref) {
+            return layers[layer.Ref].path;
+          }
+          const arnParts = layer.split(':');
+          const layerArn = arnParts.slice(0, -1).join(':');
+          const layerVersion = Number(arnParts.slice(-1)[0]);
+          const layerContentsPath = path.join(
+            '.serverless', 'layers', arnParts[6], arnParts[7]);
+          const layerContentsCachePath = path.join(
+            cachePath, 'layers', arnParts[6], arnParts[7]);
+          if (fs.existsSync(layerContentsPath)) {
+            return layerContentsPath;
+          }
+          let downloadPromise = BbPromise.resolve();
+          if (!fs.existsSync(layerContentsCachePath)) {
+            this.serverless.cli.log(`Downloading layer ${layer}...`);
+            mkdirp.sync(path.join(layerContentsCachePath));
+            downloadPromise = this.provider.request(
+              'Lambda', 'getLayerVersion', { LayerName: layerArn, VersionNumber: layerVersion })
+              .then(layerInfo => download(
+                layerInfo.Content.Location,
+                layerContentsPath,
+                { extract: true }));
+          }
+          return downloadPromise
+            .then(() => fse.copySync(layerContentsCachePath, layerContentsPath))
+            .then(() => layerContentsPath);
+        }));
+  }
+
+  buildDockerImage(layerPaths) {
+    const runtime = this.options.functionObj.runtime
```

Comment:
This piece of code is used at least 4 times in this file alone. Maybe we can use a function like `getRuntime(this)` to DRY that out a little bit...